### ISSUE TITLE
Predict results order in model.predict(source)

### ISSUE THREAD

**Issue by zjjzeng** [ultralytics/ultralytics#21500]
Hi, I found a strange case when I run my model to get inference results on a test set. First, I just used an image directory path as input source for detection. Then I changed to use related numpy array to predict again, and got different order in results. Since I tried `os.listdir ` to predict all images in the folder one by one in for-loop to validate my guess. I want to confirm will the command `model.predict(source)` probably shuffle the file order in the source path? Any help is grateful.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zjjzeng, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response, and an Ultralytics engineer will assist you here soon 😊

**Comment by Y-T-G** [comment]
It's sorted in the order of filenames:

https://github.com/ultralytics/ultralytics/blob/b905114f46311976c816d586073579a1a0b29266/ultralytics/data/loaders.py#L367

**Comment by zjjzeng** [comment]
> It's sorted in the order of filenames:
> 
> [ultralytics/ultralytics/data/loaders.py](https://github.com/ultralytics/ultralytics/blob/b905114f46311976c816d586073579a1a0b29266/ultralytics/data/loaders.py#L367)
> 
> Line 367 in [b905114](/ultralytics/ultralytics/commit/b905114f46311976c816d586073579a1a0b29266)
> 
>  files.extend(sorted(glob.glob(os.path.join(a, "*.*"))))  # dir

Thanks for your reply. I named my images like "0.png", "1.png", ... , "100.png". But I found that the prediction results were not in numerical order as expected. (compared to the numpy input or `os.listdir` for-loop). Is that normal?

**Comment by Y-T-G** [comment]
Filenames are strings. They are not going to be sorted in numerical order unless you make all of them equal digits like "001.png" instead of "1.png":

<img width="1001" height="468" alt="Image" src="https://github.com/user-attachments/assets/78614974-200b-4a8a-a798-8f0b1e7e0c21" />

**Comment by zjjzeng** [comment]
> Filenames are strings. They are not going to be sorted in numerical order unless you make all of them equal digits like "001.png" instead of "1.png":
> 
> <img alt="Image" width="1001" height="468" src="https://private-user-images.githubusercontent.com/32206511/470506972-78614974-200b-4a8a-a798-8f0b1e7e0c21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM0MDY1NDIsIm5iZiI6MTc1MzQwNjI0MiwicGF0aCI6Ii8zMjIwNjUxMS80NzA1MDY5NzItNzg2MTQ5NzQtMjAwYi00YThhLWE3OTgtOGYwYjFlN2UwYzIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI1VDAxMTcyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmNjIwNGEzN2M5YWFiM2Y5MGUwYWY1M2RlY2JjZDRjNjZkNjMwMjk2YWVlMDNiOWU5ZWZhOGNhZDQzNGI1YWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BPvd3uRU4WVdTSZTD6F-1QyZejCCHjKd37lIbwqfhGA">

Right, I just forgot this obvious thing. Thank you so much! Got it

**Comment by glenn-jocher** [comment]
You're welcome! Yes, that's exactly the expected behavior - filename sorting follows lexicographic (string) order rather than numerical order. If you need numerical ordering in the future, zero-padding your filenames like "001.png", "002.png", etc. is indeed the standard approach to ensure consistent results across different file operations.
